man/dmraid.8
.TH DMRAID 8 "DMRAID TOOL" "Heinz Mauelshagen" \" -*- nroff -*-
.SH NAME
dmraid \- discover, configure and activate software (ATA)RAID
.SH SYNOPSIS
.B dmraid
 {-a|--activate} {y|n|yes|no} 
 [-d|--debug]... [-v|--verbose]... [-i|--ignorelocking]
 [-f|--format FORMAT[,FORMAT...]]
 [-I|--ignoremonitoring]
 [{-P|--partchar} CHAR]
 [-p|--no_partitions]
 [-Z|--rm_partitions]
 [--separator SEPARATOR]
 [-t|--test]
 [RAID-set...]

.B dmraid
 {-b|--block_devices}
 [-c|--display_columns][FIELD[,FIELD...]]...
 [-d|--debug]... [-v|--verbose]...
 [--separator SEPARATOR]
 [device-path...]

.B dmraid
 {-h|--help}

.B dmraid
 {-l|--list_formats}
 [-d|--debug]... [-v|--verbose]...

.B dmraid
 {-n|--native_log}
 [-d|--debug]... [-v|--verbose]... [-i|--ignorelocking]
 [-f|--format FORMAT[,FORMAT...]]
 [--separator SEPARATOR]
 [device-path...]

.B dmraid
 {-R| --rebuild}
 RAID-set
 [device-path]

.B dmraid
 {-x| --remove}
 [RAID-set]

.B dmraid
 -f FORMAT-handler
{-C| --create} set 
 --type raidlevel
 [--size=setsize --strip stridesize] 
 --disk "device-path, device-path [, device-path ...]"

.B dmraid
[ -f|--format FORMAT-handler]
-S|--spare [RAID-set]
-M|--media  "device-path"

.B dmraid
 {-r|--raid_devices}
 [-c|--display_columns][FIELD[,FIELD...]]...
 [-d|--debug]... [-v|--verbose]... [-i|--ignorelocking]
 [-D|--dump_metadata]
 [-f|--format FORMAT[,FORMAT...]]
 [--separator SEPARATOR]
 [device-path...]

.B dmraid
 {-r|--raid_devices}
 [-d|--debug]... [-v|--verbose]... [-i|--ignorelocking]
 [-E|--erase_metadata]
 [-f|--format FORMAT[,FORMAT...]]
 [--separator SEPARATOR]
 [device-path...]

.B dmraid
 {-s|--sets}...[a|i|active|inactive]
 [-c|--display_columns][FIELD[,FIELD...]]...
 [-d|--debug]... [-v|--verbose]... [-i|--ignorelocking]
 [-f|--format FORMAT[,FORMAT...]]
 [-g|--display_group]
 [--separator SEPARATOR]
 [RAID-set...]

.B dmraid
 {-V/--version}

.SH DESCRIPTION
dmraid discovers block and software RAID devices (eg, ATARAID)
by using multiple different metadata format handlers which
support various formats (eg, Highpoint 37x series).
It offers activating RAID sets made up by 2 or more
discovered RAID devices, display properties of devices and sets (see option
.B -l
for supported metadata formats).
Block device access to activated RAID sets occurs via device-mapper nodes
/dev/mapper/RaidSetName.
RaidSetName starts with the format name (see
.B -l
option) which can be used to access all RAID sets of a specific format
easily with certain options (eg,
.B -a
below).

.SS OPTIONS
.TP
.I \-a, \-\-activate {y|n} [RAID set...]
Activates or deactivates all or particular software RAID set.
Activates or deactivates device monitoring automatically unless the
.B [-I|--ignoremonitoring]
option is given at the same time, hence preventing any dmeventd device
monitoring daemon interaction. If
.B -I
is being used on initial RAID set activation,
.B -ay
can be used later to activate device monitoring on such active RAID sets
if the RAID format supports it. See dmevent_tool(8) for device monitoring
management via this tool.
In case metadata format handlers are chosen with
.B -f
, only RAID sets with such format(s) can be activated or deactivated.
Useful if devices have multiple metadata signatures.
When activating RAID sets,
.B -p
disables the activation of partitions on them, and
.B -Z
will make dmraid tell the kernel to remove the partitions from the disks
underlying the set, ie if sda is part of the set, remove sda1, sda2, etc.
This prevents applications from directly accessiong the disks bypassing dmraid.
RAID set names given on command line don't need to be fully specified
(eg, "dmraid -ay sil" would activate all discovered Silicon Image Medley
RAID sets).

.TP
.I {-b|--block_devices} [device-path...]
List all or particular discovered block devices with their
properties (size, serial number).
Add
.B -c
to display block device names only and
.B -cc
for CSV column output of block device properties.
See description of
.B -c
below for FIELD identifiers.

.TP
.I [-d|--debug]...
Enable debugging output. Opion can be given multiple times
increasing the debug output level.

.TP
.I [-c|--display_columns][FIELD[,FIELD...]]...
Display properties of block devices, RAID sets and devices in column(s).
Optional list specifying which FIELDs to display.
.br
For
.B -b:
.br
d[evpath]|p[ath], sec[tors]|si[ze], ser[ialnumber].
.br
For
.B -r:
.br
de[vpath]|p[ath], f[ormat], r[aidname], t[ype], st[atus], se[ctors]|si[ze], da[taoffset]|o[ffset].
.br
For
.B -s:
.br
f[ormat], r[aidname], t[ype], sta[tus], str[ide], se[ctors]|si[ze], su[bsets], d[evices], sp[ares].
.br
.TP
.I [-f|--format FORMAT[,FORMAT...]]
Use metadata format handler(s) to discover RAID devices.
See
.B -l
for a list of supported format handler names. This is useful to
select particular formats in case multiple metadata signatures are found
on a device. A comma seperated list of format names can be specified which 
may not contain white space.

.TP
.I {-h|--help}
Display help text.

.TP
.I {-i|--ignorelocking}
Don't take out any locks. Useful in early boot where no read/write
access to /var is available.

.TP
.I {-I|--ignoremonitoring}
Don't activate or deactivate device monitoring with
.B -ay
or
.B -an
respectively. This avoids interaction with the dmeventd device
monitoring daemon altogether.

.TP
.I {-l|--list_formats}
List all available metadata format handlers with their names and
descriptions. Supported RAID levels are listed in parenthesis:
.br

S: Span (concatination)
.br
0: RAID0 (stripe)
.br
1: RAID1 (mirror)
.br
10: RAID10 (mirror on top of stripes) 
.br
01: RAID10 (stripe on top of mirrors) Note: Intel OROM displays this as RAID10

.TP
.I {-n|--native_log} [device-path...]
Display metadata in native, vendor-specific format.
In case a metadata format handler is chosen with
.B -f
only RAID devices with such format will be displayed in native format.
If device-path(s) is/are given on the command line, native metadata output
is restricted to those listed.
.TP
.I [{-P|--partchar} CHAR]
Use CHAR as the separator between the device name and the partition number.
.TP
.I {-R| --rebuild} RAID-set [device-path]
Rebuild raid array after a drive has failed and a new drive is added. 
For Intel chipset based systems, there are two methods in which a new drive 
is added to the system.

1. Using OROM to identify a new drive
    During system reboot, enter OROM and mark the new drive as the rebuild drive.  
    After booting to the OS, use the dmraid command to rebuild.

    Example: dmraid -R raid_set 

2. Using dmraid to identify a new drive
    Boot to the OS and use the dmraid command with the new drive as
the second parameter.

    Example: dmraid -R raid_set /dev/sdc

3. Using hot spare drive
    Mark a drive as hot spare using the "dmraid -f isw -S" command. Then use the dmraid command to start the rebuild.

    Example: dmraid -R raid_set

.TP
.I {-x|--remove} [RAID-set]
Delete one or all existing software RAID devices from the metadata.

.TP
.I -f FORMAT-handler {-C|--create} --type raidlevel [--size=setsize --strip stripsize] --disk "device-path, device-path [,device-path]"
Delete one or all existing Configure a software RAID device  and store the configuration data in a group of hard drive devices consisting of this array. This command requires the following options:

-f FORMAT-handler
.br
	metadata format (see "dmraid -l")
.br		
--type digit[digit...]
.br
	specify the raid level of the software RAID set. 
.br
		0:  raid0
.br
		1:  raid1
.br
		5:  raid5
.br
		01: raid01 (isw raid10)
.br
--size: [digits[k|K|m|M|g|G][b|B]]
.br
	specify the size of the RAID set.The number is an integer followed by [kKmMgG] and/or [bB].
.br
		b: byte (default)
.br
		B: block (512 bytes)
.br
		K or K: on the base of 1024
.br
		m or M: on the base of 1024*1024
.br
		g or G: on the base of 1024*1024*1024
.br
If this option is missing, it's set to the default value pre-configured by the vendor. Note that different vendors may apply different constraints on the granularity of the size or the minimal value.
.br
--strip: [digits[k|K|m|M|g|G][b|B]]
.br
	specify the strip size of a RAID1, RAID5, and RAID10 RAID set (as above)
.br
--disk: device-path[{,| }device-path...]
.br
	specify the array of the hard drives, e.g. /dev/sda.

.TP
.I -f FORMAT-handler -S -M device-path
.I -S -M device-path

This command adds hot spare support for one or more RAID sets.

1. When used with a format handler, which supports hot spare sets (e.g. isw), a hot spare is marked to be used when rebuilding any RAID set of that format.
2. When used when specifying a RAID set, the drive is added to that RAID set and will be used only to rebuild that set. Note: If the specified name does not match an existing RAID-set, a set with the new name will be created.

.TP
.I {-r|--raid_devices} [device-path...]
List all discovered RAID devices with format, RAID level, sectors used
and data offset into the device.
In case a metadata format handler is chosen with
.B -f
, only RAID devices with such format can be discovered. Useful if devices
have multiple metadata signatures.
If 
.B -D
is added to
.B -r
the RAID metadata gets dumped into a subdirectory named dmraid.format_name
(eg. format_name = isw) in files named devicename.dat.
The byte offset where the metadata is located on the device is written
into files named devicename.offset and the size of the device in sectors
into files named devicename.size.

If 
.B -E
is added to
.B -r
the RAID metadata on the devices gets conditionally erased.
Useful to erase old metadata after new one of different type has been
stored on a device in order to avoid discovering both. If you enter
.B -E
option
.B -D
will be enforced in order to have a fallback in case the wrong metadata
got erased.
Manual copying back onto the device is needed to recover from erasing
the wrong metadata using the dumped files devicename_formatname.dat
and devicename_formatname.offset.
Eg, to restore all *.dat files in the working directory to the respective devices:

.br
for f in *.dat
.br
do
.br
	dd if=$f of=/dev/${f%%.dat} \\
.br
	seek=`cat ${f%%dat}offset` bs=1
.br
done
.br

If device-path(s) is/are given on the command line, the above actions
are restricted to those listed.
Add
.B -c
to display RAID device names only and
.B -cc
for CSV column output of RAID device properties.
See description of
.B -c
above for FIELD identifiers.

.TP
.I --separator SEPARATOR
Use SEPARATOR as a delimiter for all options taking or displaying lists.

.TP
.I -s... [a|i] [RAID-set...]
Display properties of RAID sets. Multiple RAID set names can be given
on the command line which don't need to be fully specified (eg, "dmraid -s hpt"
would display all discovered Highpoint RAID sets). Enter
.B -s
twice to display RAID subsets too.
Add
.B -c
to display names of RAID sets only,
.B -cc
for CSV column output of RAID set properties and
.B -ccc
for inclusion of block devices in the listing. Doesn't imply
.B -s -s
to show RAID subsets (implied for group sets, e.g. isw).
Add
.B -g
to include information about group RAID sets (as with Intel Software
RAID) in the listing.
See description of
.B -c
above for FIELD identifiers.
Note: Size is given in sectors (not bytes).

.TP
.I [-v|--verbose]...
Enable verbose runtime information output. Opion can be given multiple times
increasing the verbosity level.

.SH EXAMPLES
"dmraid -l" lists all supported metadata formats with their names along with
some descriptive information, eg:
.br
hpt37x : (+) Highpoint HPT37X
.br
hpt45x : (+) Highpoint HPT45X
.br
isw    : (+) Intel Software RAID
.br
lsi    : (0) LSI Logic MegaRAID
.br
nvidia : (+) NVidia RAID
.br
pdc    : (+) Promise FastTrack
.br
sil    : (+) Silicon Image(tm) Medley(tm)
.br
via    : (+) VIA Software RAID
.br
dos    : (+) DOS partitions on SW RAIDs
.br
(0): Discover, (+): Discover+Activate

"dmraid -ay" activates all software RAID sets discovered.

"dmraid -an" deactivates all active software RAID sets which are not open
(eg, mounted filesystem on them).

"dmraid -ay -f pdc" (pdc looked up from "dmraid -l") activates all
software RAID sets with Promise format discovered and ignores all other
supported formats.

"dmraid -r" discovers all software RAID devices supported on your system, eg:
.br
/dev/dm-46: hpt45x, "hpt45x_chidjhaiaa-0", striped, ok, 320172928 sectors, data@ 0
.br
/dev/dm-50: hpt45x, "hpt45x_chidjhaiaa-0", striped, ok, 320172928 sectors, data@ 0
.br
/dev/dm-54: hpt45x, "hpt45x_chidjhaiaa-1", striped, ok, 320172928 sectors, data@ 0
.br
/dev/dm-58: hpt45x, "hpt45x_chidjhaiaa-1", striped, ok, 320172928 sectors, data@ 0


"dmraid -s -s hpt45x_chidjhaiaa" displays properties of
set "hpt45x_chidjhaiaa", eg:
.br
*** Superset
.br
name   : hpt45x_chidjhaiaa
.br
size   : 640345856
.br
stride : 128
.br
type   : raid10
.br
status : ok
.br
subsets: 2
.br
dev    : 4
.br
spare  : 0
.br
---> Subset
.br
name   : hpt45x_chidjhaiaa-0
.br
size   : 640345856
.br
stride : 128
.br
type   : stripe
.br
status : ok
.br
subsets: 0
.br
dev    : 2
.br
spare  : 0
.br
---> Subset
.br
name   : hpt45x_chidjhaiaa-1
.br
size   : 640345856
.br
stride : 128
.br
type   : stripe
.br
status : ok
.br
subsets: 0
.br
dev    : 2
.br
spare  : 0
.br

"dmraid -s -ccs hpt45" displays properties in column format of all sets
and subsets with hpt45* format, eg:
.br
hpt45x_chidjhaiaa,640345856,128,raid10,ok,4,0
.br
hpt45x_chidjhaiaa-a,640345856,128,stripe,ok,2,0
.br
hpt45x_chidjhaiaa-b,640345856,128,stripe,ok,2,0

"dmraid -r --sep : -cpath:size" display paths and sizes in sectors for
RAID devices in column format using ':' as a delimiter, eg:
.br
/dev/dm-8:320173055
.br
/dev/dm-12:320173055
.br
/dev/dm-22:320173055
.br
/dev/dm-26:320173055
.br
/dev/dm-30:586114703
.br
/dev/dm-34:586114703
.br
/dev/dm-38:586114703
.br
/dev/dm-42:586114703
.br
/dev/dm-46:156301487
.br
/dev/dm-50:156301487
.br
/dev/dm-54:390624896
.br
/dev/dm-58:390624896
.br
/dev/dm-62:390624896
.br
/dev/dm-66:390624896

"dmraid -f isw -C Raid0 --type 0 --strip 8k --size 20g --disk "/dev/sdb /dev/sdc"" creates an ISW volume with
a name of "Raid0", 20Gig bytes in total, and 8kilo bytes strip size on two disks.

"dmraid -f isw -C Test0 --type 0 --disk "/dev/sdd /dev/sde"" creates an ISW volume with the default size and strip size.

"dmraid -f isw -C Test10 --type 01 --strip 128B --disk "/dev/sda /dev/sdb /dev/sdc /dev/sdd" creates a stacked
RAID device, RAID10 (isw format), with a name of "Test10", 128 blocks (512bytes) strip size , and the default volume size on
4 disks.

"dmraid -f isw -S -M /dev/sde" marks the device /dev/sde as a hot spare for rebuild

"dmraid -R isw_djaggchdde_RAID1 /dev/sde" starts rebuild of the RAID volume on device /dev/sde

.SH DIAGNOSTICS
dmraid returns an exit code of 0 for success or 1 for error.

.SH AUTHOR
Heinz Mauelshagen <Mauelshagen@RedHat.com>


CHANGELOG

Changelog from dmraid 1.0.0.rc12 to 1.0.0.rc13		2006.10.11


FIXES:
------
o pdc.c: magic number check
o metadata.c: rd_type() returned wrong unified_type for asr



Changelog from dmraid 1.0.0.rc11 to 1.0.0.rc12		2006.09.22


FIXES:
------
o sil.c: quorate() OBO fix
o activate.c: handler() OBO fix
o added log_zero_sectors() to various metadata format handlers

FEATURES:
---------
o added SNIA DDF1 support (IBM)
o added reload functionality to devmapper.c (IBM)
o sil.[ch]: added JBOD support


MISCELANIOUS:
-------------
o streamlined devmapper.c




Changelog from dmraid 1.0.0.rc10 to 1.0.0.rc11		2006.05.16

FIXES:
------
o jm.c:	checksum() calculation

o misc.c:
	support "%d" in p_fmt and fix segfault with wrong format identifier

o nv.c:	size fix in setup_rd()

o activate.c:
	- striped devices could end on non-chunk boundaries
	- calc_region_size() calculated too small sizes causing large
	  dirty logs in memory

o isw.c: set raid5 type to left asymmetric

o toollib.c: fixed 'No RAID...' message


FEATURES:
---------

o support selection of RAID5 allocation algorithm in metadata format handlers


MISCELANIOUS:
-------------



Changelog from dmraid 1.0.0.rc9 to 1.0.0.rc10		2006.02.17

FIXES:
------
o metadata.c: want_set() didn't drop sets properly
              (eg, jmicron_raid10 wanted and jmicron_raid1 not dropped)

o scsi.c: avoid retrieving too much data (Al Viro)

o sil.h: magic number fix (some arrays were not recognized)


FEATURES:
---------
o added libdmraid_make_table() to activate.c to be used by installer

o asr.[ch]: added Adaptec HostRAID support

o jm.[ch]: added JMicron JMB36x support

o added '--enable-libselinux' to configure for those who want
  to build without it

o bytorder.h: enhanced to support big endian conversion
              on little endian arch

o nv.c: support RAID5 with dm-raid45 target

o pdc.h: support additional metadata offset

o metadata.c: enhanced to support RAID4 and RAID5 mappings with dm-raid45 target


MISCELANIOUS:
-------------
o misc.c: streamlined remove_white_space() (Al Viro)



Changelog from dmraid 1.0.0.rc8 to 1.0.0.rc9		2005.09.23

FIXES:
------
o via.c: checksum() calculation;
	 stride size

o toollib.c:  memory leak in _valid_format()

o isw.c: avoid endianess conversion bug in to_cpu();
         disk status check;
	 version check to cover 1.2.02;
	 isw_write() to store metadata in correct sequence

o hpt37x.c, lsi.c, nv.c, pdc.c, sil.c via.c: streamlined grouping switch()

o hpt45x.c: added missing RAID10 to capability string

o sil.c: streamlined quorate()

o misc.c: p_fmt() missed a free_string()


FEATURES:
---------
o pdc.c: support RAID10

o commands.c: added --separator option for selectable string
	      separator character (used with --format etc.)

o display.c: support customizable column output through field
	     identifiers with -c option

MISCELANIOUS:
------------
o metadata.c: use log_alloc_err()

o format.h: introduced caps (RAID capabilities) member

o format.c: introduced get_format_caps() and free_format_caps()
	    to ease library use in installers (pjones@redhat.com)

o display.c: streamlined log_devices()

o toollib.c: stremalined collapse_delimiters()

o isw.c: streamlined isw_read_extended() and setup_rd()

o file.c: cleanup

o add /var/lock/dmraid to rpm

o avoid global variable format_names:
  - change discover_raid_devices() interface
  - saves code in toollib

o added more info to format_error output

o activate.c: display RAID string with unsupported mapping



Changelog from dmraid 1.0.0.rc7 to 1.0.0.rc8		2005.05.19

FIXES:
------
o nv.c: fixed stripe size

o sil.c: avoid incarnation_no in name creation, because the Windows
         driver changes it every time


FEATURES:
---------
o added --ignorelocking option to avoid taking out locks in early boot
  where no read/write access to /var is given


MISCELANIOUS:
------------
o some cleanups in misc.c



Changelog from dmraid 1.0.0.rc6 to 1.0.0.rc7		2005.04.07

FIXES:
------
o pdc.c, hpt37x.c, hpt45x.c: fixed size in sectors() for RAID1


MISCELANIOUS:
------------

o sil_valid() displays the area number where the checksum is invalid now.




Changelog from dmraid 1.0.0-rc5f to 1.0.0.rc6		2005.02.28

FIXES:
------
o sil status()

o _sil_read() used LOG_NOTICE rather than LOG_INFO in order to
  avoid messages about valid metadata areas being displayed
  during "dmraid -vay".

o isw, sil filed metadata offset on "-r -D" in sectors rather than in bytes.

o isw needed dev_sort() to sort RAID devices in sets correctly.

o pdc metadata format handler name creation. Lead to
  wrong RAID set grouping logic in some configurations. 

o dos.c: partition table code fixes by Paul Moore

o _free_dev_pointers(): fixed potential OOB error

o hpt37x_check: deal with raid_disks = 1 in mirror sets

o pdc_check: status & 0x80 doesn't always show a failed device;
  removed that check for now. Status definitions needed.

o sil addition of RAID sets to global list of sets

o sil spare device memory leak

o group_set(): removal of RAID set in case of error

o hpt37x: handle total_secs > device size

o allow -p with -f

o enhanced error message by checking target type against list of
  registered target types


FEATURES:
---------

o VIA metadata format handler

o added RAID10 to lsi metadata format handler

o "dmraid -rD": file device size into {devicename}_{formatname}.size

o "dmraid -tay": pretty print multi-line tables ala "dmsetup table"

o "dmraid -l": display supported RAID levels + manual update


MISCELANIOUS:
------------

o more inline comments

o libdmraid_init() now returns lib context

o check_set() enhanced to do RAID set stack unrolling and
  to check correct number of devices in sets; saves code in
  metadata format handlers

o introduced read_raid_dev() to further reduce metadata format handler code

o optimized parse_table()

o updated dmraid manual

o devmapper.c: check target type registered before trying to load table record


tools/dmraid.c
/*
 * Copyright (C) 2004,2005  Heinz Mauelshagen, Red Hat GmbH.
 *                          All rights reserved.
 *
 * See file LICENSE at the top of this source tree for license information.
 */

/*
 * See commands.[ch] for the perform() function call abstraction below.
 */

#include <dmraid/dmraid.h>
#include "commands.h"
#include "toollib.h"
#include "version.h"

int
main(int argc, char **argv)
{
	int ret = 0;
	struct lib_context *lc;

	/* Initialize library (creates a context to use it). */
	if ((lc = libdmraid_init(argc, argv))) {
		/*
		 * Parse command line arguments and run 'early'
		 * functions for options which set library context
		 * variables (eg, --debug).
		 *
		 * Initialize locking afterwards, so that the
		 * '--ignorelocking' option can be recognized.
		 *
		 * If both are ok -> perform the required action.
		 */
		ret = handle_args(lc, argc, &argv) &&
		      init_locking(lc) &&
		      perform(lc, argv);

		/* Cleanup the library context. */
		libdmraid_exit(lc);
	}

	/* Set standard exit code. */
	exit(ret ? EXIT_SUCCESS : EXIT_FAILURE);
}
